### ISSUE TITLE
Inconsistency in Timing between Old and New Audio engine

### ISSUE THREAD

**Issue by Peter-GHub** [ppy/osu-stable-issues#447]
So as mentioned on https://github.com/ppy/osu-wiki/pull/3517 timing between 2 engines is inconsistent, when using new engine audio shifts by -15 ms so in order to match legacy engine we need to add +15 ms in the editor, all maps ranked before January 2020 were using legacy offset because new engine didn't exist while the maps after period this are either using old or new engine offset, NAT tried to make a rule in order to use either one or another engine for now it was decided that we use new engine. People who need accurate timing in order to achieve high scores and accuracy can spot the difference when switching between 2 engines (especially taiko players)

If we use new audio engine (Audio compatibility off):
- All maps between 2007- 01/2020 + those ranked with old engine timing will be shifted by 15 ms because new engine shifts audio by - 15 ms. If it will be forced on mappers then all future maps will be correctly timed to new engine however all old maps before January 2020 will be off by 15 ms with this engine

If we use old Audio engine (Audio compatibility on):
- New maps will match offset of all previous maps before January 2020 timed on old engine but it really isn't solution since we can't keep using old engine forever, currently some people actually preferred using old engine because they believed that timing will be soon fixed with new engine in order to match old engine but so far it never happened.

So best solution would be fixing offset on new engine in the way it matches old engine

It isn't problem that just I can spot because if proposal in RC was brought up to use only new engine then it's a problem that most people can spot. There were already debates about this issue that were going nowhere, not many people wants to use Github and report it.

Also reason I open this issue is so other people can discuss about it and share their experiences so we might get this fixed for now it's known issue that almost any experienced mapper knows about. some people say that they don't hear changes some people say that there is difference between gameplay and editor as well so more discussion is needed on those matters so nobody should hesitate to their share thoughts.

**Comment by peppy** [comment]
The issue here is that we did extensive testing and normalisation and the current live build matches 1:1 for all users we tested against. This makes it seem like it may be a hardware or software dependent issue.

**Comment by venix12** [comment]
Well, i can ask barely every mapper to check it and they will 99% answer it's inconsistent and almost always by exact 15ms, don't really think that's a hardware or software issues considering almost everyone gets the same result.

**Comment by pishifat** [comment]
i've asked people about this recently and found that most have the same issue as op -- timing with acm disabled results in later offsets. i also have the issue, so heres how it sounds for me: https://www.youtube.com/watch?v=Q6s9WogXJwU

results of poll asking mappers about offsets with acm on/off: https://docs.google.com/spreadsheets/d/16nOFTycmsw_BTiOJu8KUcLXcrzPgb2ZX5CdNhe0lVEo/edit#gid=1609303086
the poll asked for offsets on these 2 maps: https://osu.ppy.sh/beatmapsets/1128274#fruits/2357256 & https://osu.ppy.sh/beatmapsets/336099#osu/743896

**Comment by peppy** [comment]
have you tested with different beatmaps than those two? is it specific to certain mp3 files? further testing is appreciated if you haven’t, since any discrepancies may relate to metadata being present or not.

**Comment by pishifat** [comment]
ill test some more and come back with info soon

**Comment by peppy** [comment]
i actually have an external test app specifically used for the purpose of getting the new library in sync, will try and prepare that for distribution so we can get some more isolated testing as well.

**Comment by Kibbleru** [comment]
As of now, I've only really noticed it in the editor while checking the timing of beatmaps. I set the playback rate to 25% so I don't really think it has much to do with the framerate.

However, I can confirm that this appears very consistently in all the maps I've checked. The audio with compatibility mode off always seems to be ~10-15ms later than with compatibility mode on

**Comment by peppy** [comment]
OH MY GOD

IF YOU ARE TESTING AT 25% THEN THIS IS 100% THE ISSUE.

PLEASE, **PLEASE** only test at 100% playback.

**Comment by venix12** [comment]
Well, it's noticeably off using 100% playback anyways, using 25% is a standard for setting the offset though as it's most precise.

**Comment by peppy** [comment]
The new version is more correct at 25%. The old (legacy) code is 1000% wrong. Like it should NEVER be trusted. This is listed in most timing guides. You should ONLY time at 100%. Only. Ever. Please.

**Comment by venix12** [comment]
There never seemed to be any issues with it though, things that were set at 25% sounded pretty nice and accurately on 100% as well so I don't know.

**Comment by Peter-GHub** [comment]
15 ms difference also exists on 100% the only proof I can present for now is screenshot of recorded game soundwaves on both engines when using full speed on offset wizard
![Untitled](https://user-images.githubusercontent.com/55349441/81064294-bd1c0900-8ed9-11ea-899e-50da1b1a5c3a.png)
They are a bit different, another thing that anyone may try is to go offset wizard map and use new engine with adding +15 ms in editor and seeing if it sounds the same as on old engine on all speeds if possible

**Comment by peppy** [comment]
Could each and every person in this thread please specify software (OS) and hardware (audio card / usb interface / whatever) along with every post?

**Comment by venix12** [comment]
For what it's worth, I'm using Fiio E10k usb dac, my OS is Windows 10.

**Comment by peppy** [comment]
As a test, are you able to use onboard sound and see if anything changes?

**Comment by venix12** [comment]
Can give it a try later today I guess.

**Comment by Peter-GHub** [comment]
OS: Windows 10 1909 Build 18363.778
For hardware I'm using motherboard Asus h110m-k https://www.asus.com/Motherboards/H110M-K/specifications/ under this link may be listed everything you might need to know about my audio card I'm using built in one
CPU is I5 7400
GPU is GTX 1060
Ram is 8 GB of DDR4
Headphones are connected to jack on back of computer to motherboard

**Comment by peppy** [comment]
We found an issue which affects both lazer and stable when playback rate is not 100%. This has been fixed now, so please re-test using the latest cutting edge build and report back if this helps your scenario.

**Comment by pishifat** [comment]
works correctly for me now

**Comment by peppy** [comment]
Thanks for the confirmation. Going to close this for now, in a hope that this is the end of the issues. Feel free to continue replying.

**Comment by frukoyurdakul** [comment]
I'm going to re-open this issue as shifting the offsets in the editor isn't the actual solution to this. The gameplay is still affected with old maps and new maps being completely inconsistent, and the current fix doesn't address the issue properly. 

First things first:

I'm using a laptop and my specs are these:

1) **Graphics card**: NVIDIA GeForce GTX 1050 (2 Gb) (of course I'm using this graphics card while playing the game)

2) **Graphics card (secondary)**: Intel(R) UHD Graphics 630

3) **Processor**: Intel(R) Core(TM) i7-8750H CPU @2.20GHz (12 CPUs), ~2.2GHz

4) **DirectX version**: DirectX 12

5) **OS version**: Windows 10 Pro 64 bit

6) **osu! version**: b20200518cuttingedge (latest version as of now)

I am obviously talking about this:

![66db8145bc](https://user-images.githubusercontent.com/33460209/82737904-bcd69700-9d3c-11ea-897c-79fd02282c05.png)

I am a beatmap nominator and in the past days (including my time from almost 2 years ago) I've checked timings with both excessive testplays and checks from editor multiple times, with 100% speed in almost all cases. Those maps' offsets are also different, I am still getting at least +12 ms difference while playing with new engine compared to the old one. This also applies to the new maps where I get different results if I activate the compatibility mode. Fixing the issues on the editor doesn't seem to be helping in this case I'm afraid, they are patches that letting people use 25% speed or 50% while timing the maps, but it's not how it's supposed to be done anyways because music distorts and that itself is another problem.

Going back to the topic, I am going to share 2 screenshots from the same map of the same section (well, not exactly the same section but you'll get the point) where even I have tried to hit them in time with the music, I got noticeably different results.

New engine:

![image](https://user-images.githubusercontent.com/33460209/82737970-28b8ff80-9d3d-11ea-8b24-a48102de95bb.png)

Old engine: 

![image](https://user-images.githubusercontent.com/33460209/82737981-3ec6c000-9d3d-11ea-95e7-27c0d80f7375.png)

On the UR bar at the bottom, you can clearly see I am hitting late unintentionally in the first screenshot where I hear the music is perfectly fine with my playing on both of them. I am also as accurate as the first one in the second picture, where in the UR bar the hits are perfectly in the middle, even a bit earlier which might be from my local offset itself.

The latest fix did not help at all. The issue persists, and it is affecting gameplay, if not editor itself. To be honest, the 25% rate or 50% rate speeds should be fixed because editor itself is shifting the offset, but that's another case. The maps themselves are still affected in the gameplay.

I believe the issue itself is happening because of how the new engine is more efficient than the old one, and the lag on either input or output of the audio between both engines is causing this difference.

I can think of a solution (or, another patch) at this point: Depending on when exactly this issue started to appear, the **gameplay engine** should shift the offset itself by +15 ms if the audio compatibility mode is off and the latest update of the map is earlier than the issue start date. Both informations should be present in the game.

The +15 ms value can also be discussed further but it was already been discussed by several people from google forms and discussions on the BNG server.

Bottom line: The new engine and the old engine aren't synchronized enough because of the effectiveness of the new one compared to the old one which reduces lag, relatively causing the offset issues on older maps.

**Comment by peppy** [comment]
That sounds like you need to adjust your UO accordingly, which should fix your issue. Have you tried this?

**Comment by frukoyurdakul** [comment]
Universal offset is a different thing though. Both plays have done with +2ms universal offset which is my input lag. But there are noticeable differences on both of the screen shots which describes the actual issue, which only happens by turning on and off audio compatibility mode. So I am certain that is the matter that causes the offset difference.

**Comment by peppy** [comment]
It's not a different thing. UO is made to offset audio latency in music playback, not input (although obviously many use it for input as well). Changing to the new engine may be causing an issue with your audio drivers or hardware which changes this offset. UO may need adjusting. So please try adjusting UO and report back if you have issues.

Case in point: if you can fix the new engine being incorrect by using UO (the same UO on all maps) you are facing a slightly different issue to the one mentioned in this thread, which has been confirmed fixed by multiple people after the recent release.
